Core/GameEngine/Source/Common/System/MiniDumper.cpp

```diff
@@ -0,0 +1,698 @@
+	switch (m_dumpObjectsState)
+	{
+	case MEMORY_POOLS:
+	{
+		// Dump all the MemoryPool instances in TheMemoryPoolFactory
+		// This only dumps the metadata, not the actual MemoryPool contents (done in the next phase).
+		if (TheMemoryPoolFactory == NULL)
+		{
+			m_dumpObjectsState = MEMORY_POOL_ALLOCATIONS;
+			break;
+		}
+
+		Int poolCount = TheMemoryPoolFactory->getMemoryPoolCount();
+		//m_dumpObjectsSubState contains the index in TheMemoryPoolFactory of the MemoryPool that is being processed
+		if (m_dumpObjectsSubState < poolCount)
+		{
+			MemoryPool* pool = TheMemoryPoolFactory->getMemoryPoolN(m_dumpObjectsSubState);
+			if (pool != NULL)
+			{
+				memoryBase = reinterpret_cast<ULONG64>(pool);
+				memorySize = sizeof(MemoryPool);
+				++m_dumpObjectsSubState;
+			}
+			else
+			{
+				m_dumpObjectsSubState = poolCount;
+			}
+		}
+
+		if (m_dumpObjectsSubState == poolCount)
+		{
+			m_dumpObjectsSubState = 0;
+			m_dumpObjectsState = MEMORY_POOL_ALLOCATIONS;
+		}
+		break;
+	}
+	case MEMORY_POOL_ALLOCATIONS:
+	{
+		// Iterate through all the allocations of memory pools and containing blobs that has been done via the memory pool factory
+		// and include all of the storage space allocated for objects
+		if (TheMemoryPoolFactory == NULL)
+		{
+			m_dumpObjectsState = DMA_ALLOCATIONS;
+			break;
+		}
+
+		//m_dumpObjectsSubState is used to track if the iterator needs to be initialized, otherwise just a counter of the number of items dumped
+		if (m_dumpObjectsSubState == 0)
+		{
+			m_rangeIter = TheMemoryPoolFactory->cbegin();
+			++m_dumpObjectsSubState;
+		}
+
+		// m_RangeIter should != cend() at this point before advancing, unless the memory pool factory is corrupted (or has 0 entries)
+		memoryBase = reinterpret_cast<ULONG64>(m_rangeIter->allocationAddr);
+		memorySize = m_rangeIter->allocationSize;
+		++m_dumpObjectsSubState;
+		++m_rangeIter;
+
+		if (m_rangeIter == TheMemoryPoolFactory->cend())
+		{
+			m_dumpObjectsState = DMA_ALLOCATIONS;
+			m_dumpObjectsSubState = 0;
+		}
+		break;
+	}
+	case DMA_ALLOCATIONS:
+	{
+		// Iterate through all the direct allocations ("raw blocks") done by DMAs, as these are done outside of the
+		// memory pool factory allocations dumped in the previous phase.
+		if (TheDynamicMemoryAllocator == NULL)
+		{
+			m_dumpObjectsState = COMPLETED;
+			break;
+		}
+
+		if (m_currentAllocator == NULL)
+		{
+			m_currentAllocator = TheDynamicMemoryAllocator;
+			// m_dumpObjectsSubState is used to track the index of the raw block in the allocator we are currently traversing
+			m_dumpObjectsSubState = 0;
+		}
+
+		MemoryPoolAllocatedRange rawBlockRange = {0};
+		int rawBlocksInDma = m_currentAllocator->getRawBlockCount();
+		if (m_dumpObjectsSubState < rawBlocksInDma)
+		{
+			// Dump this block
+			m_currentAllocator->fillAllocationRangeForRawBlockN(m_dumpObjectsSubState, rawBlockRange);
```

Comment:
`fillAllocationRangeForRawBlockN` also has a loop inside.

Can we perhaps cache the MemoryPoolSingleBlock* pointer and just keep going to next one until it is NULL, without using `getRawBlockCount`, `fillAllocationRangeForRawBlockN`?